portfolio/src/main/java/com/google/sps/servlets/RecommendBookServlet.java

```diff
@@ -0,0 +1,47 @@
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+import com.google.sps.data.Book;
+import com.google.sps.data.BookReader;
+import com.google.sps.data.GenreRecommender;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/bookRecommendationById")
+public class RecommendBookServlet extends HttpServlet {
+  Map<Integer, Book> bookList;
+  @Override
+  public void init() throws ServletException {
+    try {
+      BookReader reader = new BookReader(getServletContext().getRealPath("/WEB-INF/20_books.csv"));
+      bookList = reader.makeBookList();
+    } catch (Exception ex) {
+      throw new ServletException("Error reading CSV file", ex);
+    }
+  }
+  private static String toJSON(List<Book> books) {
+    Gson gson = new Gson();
+    return gson.toJson(books);
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int bookId = Integer.parseInt(request.getParameter("id"));
+    Book book = bookList.get(bookId);
+    List<Book> fullBookList = new ArrayList<>(bookList.values());
+    GenreRecommender rec = new GenreRecommender(fullBookList);
```

Comment:
Sorry, one more nit I didn't see before:
we can make this a little more concise by getting rid of fullBookList and just having "new GenreRecommender(bookList.values))"

Generally, I try to avoid creating new variables unless they are going to be used multiple times (although it does sometimes help readability)

Reply:
Makes sense! I still had to make a new ArrayList object though since .values() returns a Collection (which apparently can't automatically be converted to a List)